### PR TITLE
Implement VirtualFile for preview file diffs 

### PR DESCRIPTION
**Pull request by eseliger**
This was one open todo for myself, leftover from #9166 

Due to the simplifications, this is also addressing the concerns from https://github.com/sourcegraph/sourcegraph/issues/10616

Closes https://github.com/sourcegraph/sourcegraph/issues/10616

### CODE REVIEW

**Comment by eseliger** [comment]
@mrnugget was able to get some more stuff cleaned up.

Next tasks would be tests IMO.
- [ ] FileDiffResolver.Highlight has none
- [ ] Pagination tests in campaigns/resolver_test.go should be moved to repo_comparison_test.go (doesn't exist yet)
- [ ] CodeLines in package highlight doesn't have any tests

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -2646,6 +2609,41 @@ interface File2 {
     ): HighlightedFile!
 }
 
+# An external file is an arbitrary file that is not stored in git.
+type ExternalFile implements File2 {
```
Comment (felixfbecker) [inline cmd/frontend/graphqlbackend/schema.graphql:2613]: I feel like "external" doesn't match the description - "external" with changesets refers to pull requests stored on code hosts, but that's not the case here (?). Maybe "VirtualFile" would be a better fit?

Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql:2613]: We talked about that and had that name on our list as well. `ExternalFile` is supposed to be a super generic one, and at some point, when we have a blob store, we'd like to store that filecontent there as well (and logs, extension bundles, ..) 
A not so hard (but not DRY at all) way would be to implement two equal files, one being `ExternalFile` (the generic blob store thingy), and another `VirtualFile`, for files that don't exist anywhere on disk _at all_. 
I'll think about it

Comment (felixfbecker) [reply cmd/frontend/graphqlbackend/schema.graphql:2613]: Hmm, I could imagine the blob store file will eventually also have some specific properties, so having separate types makes sense to me. I think DRY is an unachievable goal in GraphQL IDL anyway (just think about all the Connections...)

File: cmd/frontend/graphqlbackend/repository_comparison.go [hunk 2]
```diff
@@ -624,7 +624,3 @@ func (r *DiffStat) AddDiffStat(s *DiffStat) {
 func (r *DiffStat) Added() int32   { return r.added }
 func (r *DiffStat) Changed() int32 { return r.changed }
 func (r *DiffStat) Deleted() int32 { return r.deleted }
-
-func (r *DiffStat) String() string {
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/repository_comparison.go]: Why did you remove that?

File: cmd/frontend/graphqlbackend/repository_comparison_test.go [hunk 3]
```diff
@@ -499,12 +501,17 @@ func TestFileDiffHighlighter(t *testing.T) {
 	ctx := context.Background()
 
 	file1 := &dummyFileResolver{
-		content: func(ctx context.Context) (string, error) { return "foobar", nil },
+		content: func(ctx context.Context) (string, error) { return "foobar\nbarfoo\ndone\n", nil },
 	}
 	file2 := &dummyFileResolver{
-		content: func(ctx context.Context) (string, error) { return "foobar", nil },
+		content: func(ctx context.Context) (string, error) { return "foobar\nbarfoo\ndone\n", nil },
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/repository_comparison_test.go:507]: I think it would be nice if the two files had different content in the tests and that a reader can then see the expected behavior in the rest of the assertions. Right now the code returned by `highlight.Mocks.Code` looks like it could be from `file1` or `file2`.

The assertions for `highlightedBase` and highlightedHead` look the same.

You could return different content in `highlight.Mocks.Code` by checking the `Path` of the `highlight.Params`.

File: cmd/frontend/graphqlbackend/repository_comparison_test.go [hunk 4]
```diff
@@ -499,12 +501,17 @@ func TestFileDiffHighlighter(t *testing.T) {
...
 	}
 
+	highlight.Mocks.Code = func(p highlight.Params) (template.HTML, bool, error) {
+		return "<table><tbody><tr><td class=\"line\" data-line=\"1\"></td><td class=\"code\"><div><span style=\"color:#657b83;\">foobar\n</span></div></td></tr><tr><td class=\"line\" data-line=\"2\"></td><td class=\"code\"><div><span style=\"color:#657b83;\">barfoo\n</span></div></td></tr><tr><td class=\"line\" data-line=\"3\"></td><td class=\"code\"><div><span style=\"color:#657b83;\">done</span></div></td></tr></tbody></table>", false, nil
+	}
+	defer func() { highlight.Mocks.Code = nil }()
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/repository_comparison_test.go]: I changed this to use your `ResetMocks` function.

File: enterprise/internal/campaigns/resolvers/resolver_test.go [hunk 5]
```diff
@@ -1474,6 +1474,268 @@ func TestCreateCampaignWithPatchSet(t *testing.T) {
...
+		}
+		return ioutil.NopCloser(strings.NewReader(testDiff)), nil
+	}
+	defer func() { git.Mocks.ExecReader = nil }()
+
+	git.Mocks.ResolveRevision = func(spec string, opt *git.ResolveRevisionOptions) (api.CommitID, error) {
+		return api.CommitID(testBaseRevision), nil
+	}
+	defer func() { git.Mocks.ResolveRevision = nil }()
+
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/resolver_test.go:1580]: Do we need all of these mocks and the same `testDiff` with 3 files?

File: cmd/frontend/graphqlbackend/external_file.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	return "", nil
+}
+
+func (r *externalFileResolver) CanonicalURL() (string, error) {
+	// Todo: allow viewing arbitrary files in the webapp.
+	return "", nil
+}
+
+func (r *externalFileResolver) ExternalURLs(ctx context.Context) ([]*externallink.Resolver, error) {
+	// Todo: allow viewing arbitrary files in the webapp.
```
[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/graphqlbackend/external_file.go]: Will do at some point, not needed now and documented in the schema. They are easy to distinguish using `__typename`

**Comment by eseliger** [comment]
@slimsag I carefully checked it should address all your comments on the issue, can you please check if the changes make sense to you?

File: cmd/frontend/graphqlbackend/git_tree_entry_test.go [hunk 7]
```diff
@@ -20,3 +24,35 @@ func TestGitTreeEntry_RawZipArchiveURL(t *testing.T) {
 		t.Errorf("got %q, want %q", got, want)
 	}
 }
+
+func TestGitTreeEntry_Content(t *testing.T) {
+	testPath := "foobar.md"
+	testContent := "foobar"
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/git_tree_entry_test.go:30]: nit:
```suggestion
	wantPath := "foobar.md"
	wantContent := "foobar"
```

File: cmd/frontend/graphqlbackend/git_tree_entry_test.go [hunk 8]
```diff
@@ -20,3 +24,35 @@ func TestGitTreeEntry_RawZipArchiveURL(t *testing.T) {
...
+
+	git.Mocks.ReadFile = func(commit api.CommitID, name string) ([]byte, error) {
+		if name != testPath {
+			t.Fatalf("wrong name in ReadFile call. want=%q, have=%q", testPath, name)
+		}
+		return []byte(testContent), nil
+	}
+
+	defer func() { git.Mocks.ReadFile = nil }()
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/git_tree_entry_test.go:39]:
```suggestion
	t.Cleanup(func() { git.Mocks.ReadFile = nil })
```

File: cmd/frontend/graphqlbackend/repository_comparison_test.go [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+	}
+
+	git.Mocks.GetCommit = func(id api.CommitID) (*git.Commit, error) {
+		if string(id) != testBaseRevision && string(id) != testHeadRevision {
+			t.Fatalf("GetCommit received wrong ID: %s", id)
+		}
+
+		return &git.Commit{ID: id}, nil
+	}
+	defer func() { git.Mocks.GetCommit = nil }()
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/repository_comparison_test.go:49]:
```suggestion
	t.Cleanup(func() { git.Mocks.GetCommit = nil })
```

Same for other occurrences.

File: cmd/frontend/graphqlbackend/repository_comparison_test.go [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/gitserver"
+	"github.com/sourcegraph/sourcegraph/internal/highlight"
+	"github.com/sourcegraph/sourcegraph/internal/vcs/git"
+)
+
+func TestRepositoryComparison(t *testing.T) {
+	ctx := context.Background()
+
+	testBaseRevision := "24f7ca7c1190835519e261d7eefa09df55ceea4f"
+	testHeadRevision := "b69072d5f687b31b9f6ae3ceafdc24c259c4b9ec"
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/repository_comparison_test.go:28]: nit:
```suggestion
	wantBaseRevision := "24f7ca7c1190835519e261d7eefa09df55ceea4f"
	wantHeadRevision := "b69072d5f687b31b9f6ae3ceafdc24c259c4b9ec"
```

File: enterprise/internal/campaigns/resolvers/patch_sets_test.go [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
++Another line
+`
+
+	now := time.Now().UTC().Truncate(time.Microsecond)
+	clock := func() time.Time {
+		return now.UTC().Truncate(time.Microsecond)
+	}
+
+	testBaseRevision := "24f7ca7c1190835519e261d7eefa09df55ceea4f"
+	testHeadRevision := "b69072d5f687b31b9f6ae3ceafdc24c259c4b9ec"
```
Comment (unknwon) [inline enterprise/internal/campaigns/resolvers/patch_sets_test.go:103]: nit:
```suggestion
	wantBaseRevision := "24f7ca7c1190835519e261d7eefa09df55ceea4f"
	wantHeadRevision := "b69072d5f687b31b9f6ae3ceafdc24c259c4b9ec"
```

File: enterprise/internal/campaigns/resolvers/patch_sets_test.go [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+				t.Fatal(err)
+			}
+
+			if pageInfo.HasNextPage() != tc.wantHasNextPage {
+				t.Fatalf("pageInfo HasNextPage wrong. want=%t, have=%t", tc.wantHasNextPage, pageInfo.HasNextPage())
+			}
+
+			if diff := cmp.Diff(pageInfo.EndCursor(), tc.wantEndCursor); diff != "" {
+				t.Fatal(diff)
+			}
```
Comment (unknwon) [inline enterprise/internal/campaigns/resolvers/patch_sets_test.go:214]: IIRC our current convention is to put "want" as first argument:
```suggestion
			if diff := cmp.Diff(tc.wantEndCursor, pageInfo.EndCursor()); diff != "" {
				t.Fatalf("(-want +got):\n%s", diff)
			}
```

Same for other occurrences.

File: cmd/frontend/graphqlbackend/virtual_file.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	return highlight.IsBinary([]byte(content)), nil
+}
+
+func (r *virtualFileResolver) Highlight(ctx context.Context, args *HighlightArgs) (*highlightedFileResolver, error) {
+	content, err := r.Content(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return highlightContent(ctx, args, content, r.Path(), highlight.Metadata{
+		Revision: fmt.Sprintf("Preview file diff %s", r.stat.Name()),
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/virtual_file.go:78]: Are you certain we cannot also pass the repository name here? Although it is optional, without it we basically cannot determine where the file came from when highlighting fails and as such have no way of debugging why highlighting failed on the file

File: internal/highlight/highlight.go [hunk 14]
```diff
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
...
 }
 
-// ParseLinesFromHighlight takes the highlighted html table and returns a slice of highlighted strings, where each string corresponds a single line in the original, highlighed file.
-func ParseLinesFromHighlight(input string) ([]string, error) {
-	doc, err := html.Parse(strings.NewReader(input))
+// Lines highlights the file and returns a list of highlighted lines.
+// The returned boolean represents whether or not highlighting was aborted due
+// to timeout.
+func Lines(ctx context.Context, p Params) ([]string, bool, error) {
+
```
Comment (slimsag) [inline internal/highlight/highlight.go:463]:
```suggestion
```

File: internal/highlight/highlight.go [hunk 15]
```diff
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
 	return buf.String(), nil
 }
 
-// ParseLinesFromHighlight takes the highlighted html table and returns a slice of highlighted strings, where each string corresponds a single line in the original, highlighed file.
-func ParseLinesFromHighlight(input string) ([]string, error) {
-	doc, err := html.Parse(strings.NewReader(input))
+// Lines highlights the file and returns a list of highlighted lines.
+// The returned boolean represents whether or not highlighting was aborted due
+// to timeout.
+func Lines(ctx context.Context, p Params) ([]string, bool, error) {
```
Comment (slimsag) [inline internal/highlight/highlight.go:462]: This appears to be returning HTML, so I think you should be returning `[]template.HTML` and not `[]string` here to make that clear.

Comment (slimsag) [inline internal/highlight/highlight.go:462]:
```suggestion
func CodeAsLines(ctx context.Context, p Params) ([]string, bool, error) {
```

I can imagine someone being confused that this requires the while file and "not just a few lines", so I suggest renaming to `CodeAsLines` instead.

File: internal/highlight/highlight.go [hunk 16]
```diff
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
...
+
+	html, aborted, err := Code(ctx, p)
+	if err != nil {
+		if err == errBinary {
+			return nil, aborted, nil
```
Comment (slimsag) [inline internal/highlight/highlight.go:467]: This means if the input was binary and could not be highlighted at all, you return no HTML *and* no error. That seems a bad API to me. I think you should return the error and handle it properly at the call sites instead.

Comment (slimsag) [reply internal/highlight/highlight.go:467]: Additionally, get rid of `errBinary` once you do this because it becomes useless again.

File: internal/highlight/highlight.go [hunk 17]
```diff
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
...
+		}
+		return nil, aborted, err
+	}
+	lines, err := splitHighlightedLines(html)
+	return lines, aborted, err
+}
+
+// splitHighlightedLines takes the highlighted HTML table and returns a slice
+// of highlighted strings, where each string corresponds a single line in the
+// original, highlighed file.
```
Comment (slimsag) [inline internal/highlight/highlight.go:477]:
```suggestion
// original, highlighted file.
```

File: internal/highlight/highlight.go [hunk 18]
```diff
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
...
+
+	html, aborted, err := Code(ctx, p)
+	if err != nil {
+		if err == errBinary {
+			return nil, aborted, nil
+		}
+		return nil, aborted, err
+	}
+	lines, err := splitHighlightedLines(html)
+	return lines, aborted, err
```
Comment (slimsag) [inline internal/highlight/highlight.go:472]: Also please add a basic test for this function (nothing fancy, just one for future proofing)

File: internal/highlight/highlight.go (hunk 16 above)
Comment (eseliger) [reply internal/highlight/highlight.go:466]: It was meant to address your concern in the issue:

> Binary-file checks like this, this, and this appear to be sprinkled throughout the code. This smells to me of defensive coding, especially because highlight.Code already handles binary files appropriately. These checks should not be needed anywhere I believe, because if the file is binary highlight.Code is going to give an error which should be handled without a prior "is binary" check.

We _do_ allow binary renders to fail, as we still might have a non-binary head file for example, and hence it would be just additions of the head in the final output.
I think returning `[]string{}` should be good here instead to make that intention more clear? Then, we wouldn't have to expose the Error so it can be checked on in referencing code. Otherwise, we can get rid of it, but would have to keep the error around, for assertions

File: internal/highlight/highlight.go [hunk 19]
```diff
@@ -85,13 +85,19 @@ type Metadata struct {
...
 
+// errBinary is returned when a binary file was attempted to be highlighted.
+var errBinary = errors.New("cannot render binary file")
+
 // Code highlights the given file content with the given filepath (must contain
 // at least the file name + extension) and returns the properly escaped HTML
 // table representing the highlighted code.
 //
 // The returned boolean represents whether or not highlighting was aborted due
 // to timeout. In this scenario, a plain text table is returned.
```
Comment (slimsag) [inline internal/highlight/highlight.go:96]:
```suggestion
// to timeout. In this scenario, a plain text table is returned.
//
// In the event the input content is binary, an error is returned.
```

File: internal/highlight/highlight.go [hunk 20]
```diff
@@ -450,14 +456,31 @@ func unhighlightLongLines(h string, n int) (string, error) {
 	return buf.String(), nil
 }
 
-// ParseLinesFromHighlight takes the highlighted html table and returns a slice of highlighted strings, where each string corresponds a single line in the original, highlighed file.
-func ParseLinesFromHighlight(input string) ([]string, error) {
-	doc, err := html.Parse(strings.NewReader(input))
+// CodeAsLines highlights the file and returns a list of highlighted lines.
+// The returned boolean represents whether or not highlighting was aborted due
+// to timeout. Returns no lines for binary files.
```
Comment (slimsag) [inline internal/highlight/highlight.go:461]:
```suggestion
// to timeout.
//
// In the event the input content is binary, ErrBinary is returned.
```

File: internal/highlight/highlight.go (hunk 16 above)
Comment (slimsag) [reply internal/highlight/highlight.go]: The concern of mine you quoted was primarily about the fact that we checked if the file was binary _before_ passing it to `highlight.Foobar`, instead of relying on `highlight.Foobar` to detect this and tell us it is binary. Also, it was about the fact that we did this in several places.

As I understand it now, that problem is mostly addressed and you can just expose `ErrBinary` now and check it once here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@es/improve-filediffs/-/blob/cmd/frontend/graphqlbackend/repository_comparison.go#L482-488

I think that is OK and desirable because, logically, if you try to highlight a binary file for display you would expect to get an error and not simply "no result" (empty lines).

In other words, empty lines make sense for your use case of `highlight.CodeAsLines` but not other potential use cases, and thus we should keep that logic close to your use case. Do you buy that argument?

Comment (eseliger) [reply internal/highlight/highlight.go]: absolutely! adjusting it now

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 21]
```diff
@@ -2636,6 +2599,41 @@ interface File2 {
...
+    # rich file type.
+    #
+    # This HTML string is already escaped and thus is always safe to render.
+    richHTML: String!
+    # Not implemented.
+    url: String!
+    # Not implemented.
+    canonicalURL: String!
+    # Not implemented.
+    externalURLs: [ExternalLink!]!
```
Comment (felixfbecker) [inline cmd/frontend/graphqlbackend/schema.graphql:2624]: I don't quite see the value of introducing this type that implements this interface, if it doesn't actually implement the interface. That actually seems harmful to me, because now in TypeScript we have types that are non-nullable per schema but then contain an empty string, which could lead to nasty bugs (it has in the past with similar cases, e.g. #5125 is one example of many).

We may say "this is just temporary until we have pages for these so it's fine" but experience shows that these "temporary" things stay in there longer than expected because adding the UI doesn't become a product priority and they become _yet another_ wart of our GraphQL API, just like

https://github.com/sourcegraph/sourcegraph/blob/09f5fd308e10e0a8fa15fd73a9a71fa7b1ec98af/cmd/frontend/graphqlbackend/schema.graphql#L2773

and I would really like to reduce the number of those warts, not increase them.

The alternative, proper way would be to say `File`s do not necessarily need to have a URL, because not all implementations can provide a URL. Only `GitBlob`s would have the `url` fields. I think that is sensible. A variety of GraphQL types will then have to be updated of course, for example `repositoryComparison` uses a `FileDiffConnection` that references `File`s, but really it should be a `GitBlobDiffConnection` since comparing two git revisions will always yield diffs between git blobs, never virtual files, and this would force us to correct that because we need the `url` field in the UI.

That would of course increase the number of types in our schema, but DRY is an unachievable goal in GraphQL IDL due to lack of generics. `Connection` is an inherently generic type, that, in IDL, must be "expanded" into every instance. In my eyes it's the price to pay for the type safety and tooling GraphQL brings (at least if you want to use IDL vs the server declaring the schema, but that is another topic). We already have 28 connection types, having a few more  doesn't make a difference in my eyes.

Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql:2624]: > The alternative, proper way would be to say Files do not necessarily need to have a URL, because not all implementations can provide a URL. Only GitBlobs would have the url fields. I think that is sensible. A variety of GraphQL types will then have to be updated of course, for example repositoryComparison uses a FileDiffConnection that references Files, but really it should be a GitBlobDiffConnection since comparing two git revisions will always yield diffs between git blobs, never virtual files, and this would force us to correct that because we need the url field in the UI.

Exactly that is not the case haha, why we introduced the VirtualFile here is so we _can_ have a proper comparison between the two, for diffs that don't yet exist on gitserver. 

I am strongly against having to have two kinds of repo comparisons again, one for git comparisons and one for virtual file comparisons. This mess was just cleaned up here and I don't think it is something that should exist in the codebase, as it is veery duplicative, if we can get around it in any way. 

However, the decision to put URL,.. on File2, not GitBlob might have been a wrong one. Anyways, that would be a breaking change and would need some releases until we can finally do it. If we decide to go that way, we could still keep the `File2` on the repo comparison, though, so we don't have to duplicate all that stuff:

```graphql
...
oldFile {
  __typename
  path
  name
  ... on GitBlob {
    url
}
```

I think the reasonable approach would be to slowly deprecate URL,.. on File2 
 and move it over to `GitBlob`. WDYT?
